### ISSUE TITLE
fmin/fmax semantics are problematic when comparing Jets with scalars

### ISSUE THREAD

**Issue by sandwichmaker** [ceres-solver/ceres-solver#816]
cc: @sergiud 

@alexsmac found an instance of this in production code. But here is a minimal reproduction. Consider the following two tests.

```
TEST(Jet, FmaxEqualRealPartLeft) {
  const J x = MakeJet(1, 2, 3);
  const J y = fmax(x, x.a);
  EXPECT_EQ(y.a, x.a) << y << " " << x;
  EXPECT_EQ(y.v, x.v) << y << " " << x;
}

TEST(Jet, FmaxEqualRealPartRight) {
  const J x = MakeJet(1, 2, 3);
  const J y = fmax(x.a, x);
  EXPECT_EQ(y.a, x.a) << y << " " << x;
  EXPECT_EQ(y.v, x.v) << y << " " << x;
}

```

```
[ RUN      ] Jet.FmaxEqualRealPartLeft
/Users/sameeragarwal/ceres-solver/internal/ceres/jet_test.cc:1295: Failure
Expected equality of these values:
  y.v
    Which is: { 0, 0 }
  x.v
    Which is: { 2, 3 }
[1 ; 0, 0] [1 ; 2, 3]
[  FAILED  ] Jet.FmaxEqualRealPartLeft (0 ms)
[ RUN      ] Jet.FmaxEqualRealPartRight
[       OK ] Jet.FmaxEqualRealPartRight (0 ms)
[----------] 34 tests from Jet (0 ms total)
```

`FmaxEqualPartRight` passes while `FmaxEqualPartLeft` fails. The issue is the way we do jet v/s scalar comparisons. We promote the scalar to a jet with zero derivative and compare the real parts. If both the operands are jets to begin with then what derivative is returned is ambiguous. 

However if one of them is scalar, then depending on whether the scalar is first operand or second, we can end up with a completely zero derivative, which does not seem ideal. The same problem is present with `fmin`.

**Comment by sergiud** [comment]
On equality the argument with non-zero dual part (alternatively, the `Jet` argument) should be prioritized, correct?

**Comment by sergiud** [comment]
I'm also not convinced treating equality in a special manner is the way to go. As another possibility:  what if the user expects a zero derivative? This case would be always overridden by the proposed logic.

Even if we try to be smart here, as you mentioned, the problem only shifts to Jet-Jet comparisons.

I believe establishing a convention is more sensible. This way, the user can decide which argument is prioritized in terms of the order the arguments are passed to both functions.

**Comment by sandwichmaker** [comment]
lets deal with the two issues one at a time. 

What do you think is a reasonable solution for scalar/jet asymmetry?

as for establishing a convention, I am not sure that will do much. Because libraries meant for automatic differentiation are written without being aware of ceres for one, the other is that it is possible that the author may not have thought about fmax/fmin at all. 

I think if we have an ambiguous situation, then we should make choices that minimize surprises. 
I think in that case, I think @alexsmac is right that the less surprising thing would to prioritize the jet over the scalar value  when the real part of the jet is equal to the scalar. This will also resolve the asymmetry.

It leaves the ambiguous comparison of two jets as the mess it already is.

**Comment by sergiud** [comment]
As previously indicated, I do not believe these are two separate problems. Treating the issues separately would clearly cause an inconsistency in the produced results. Eventually, we want to be consistent to ternary comparison (using ordinary comparison operators) that replicates `fmin/fmax`.

**Comment by sandwichmaker** [comment]
Sorry that was a bit of a rambling comment, and  while I started by saying that we should separate them I ended up elsewhere. Let me try and re-state my position.

When there is ambiguity, we should try and minimize surprise. I think the in the case of comparing a scalar and a jet, the thing which minimizes surprises is:

1. The result should be the same regardless of whether it is fmax(s,j) or fmax(j,s).
2. if j.a == s then the result is j.

This does not solve the issue of fmax(j1, j2) where j1.a == j2.a, but there again while it seems clear to me that we should have the same answer regardless of the order of the arguments to fmax, it is not clear to me what that answer should be.

**Comment by sergiud** [comment]
Thanks for clarification. Unfortunately, this still leaves my question unanswered (given your intention is to minimize surprises) why `fmin/fmax` should behave differently from
```cpp
a > b ? a : b // fmax
a < b ? a : b // fmin
```
where `a`, `b` are combinations of arguments supported by both functions.

**Comment by sandwichmaker** [comment]
My argument (albeit not a particularly strong one) is that when a user uses fmax/fmax they are not thinking in terms of an implementation.  They are are primarily thinking in terms of an implementation. Indeed they would likely be perfectly happy with 

```
a >= b ? a : b // fmax
a <= b ? a : b // fmin
```

**Comment by sergiud** [comment]
I see. However, `fmin/fmax` are essentially the above comparisons (pick any of the two) sans floating point exceptions. I still argue there should not be any inconsistency between any of the variants (as a matter of fact, neither compared to `std::min` and `std::max` results.)

**Comment by sergiud** [comment]
As before: establishing a convention such that users can decide what derivatives are propagated on equality. I do not believe we can do better that without sacrificing consistency.

**Comment by sandwichmaker** [comment]
No amount of convention setting is going to get around the surprise of `fmax(s,j) != fmax(j,s)`.

**Comment by sergiud** [comment]
Sure, but the alternative is equally surprising for Jet arguments, no?

**Comment by alexsmac** [comment]
I'm not a fan of any solution that requires custom functions such as `Propagate()` as it is pretty common amongst our user-base for people to use Jets implicitly in generic and/or 3rd party templated math libraries, that's a key USP and requiring those to use custom Ceres specific functions and thus link against Ceres is not viable solution.

I think the biggest issue here is that whilst the commit message for the change implied a change purely to cover NANs, we actually changed the behaviour for finite values too, and that is an unwelcome surprise.

Prior to the commit when we used `operator<` throughout (even in `fmax()`) we had:

```
template <typename T, int N>
inline Jet<T, N> fmax(const Jet<T, N>& x, const T& y) {
  return x < y ? Jet<T, N>{y} : x;
}
template <typename T, int N>
inline Jet<T, N> fmax(const T& x, const Jet<T, N>& y) {
  return x < y ? y : Jet<T, N>{x};
}
```
For the case where x == y (as returned by our comparators, i.e. one value is a scalar, one a Jet and the real part of the Jet and the scalar are equal, then:
```
fmax(const Jet<T, N>& x, const T& y) -> x
fmax(const T& x, const Jet<T, N>& y) -> Jet<T,N>{x}
```
After this commit, we have:
```
fmax(const Jet<T, N>& x, const T& y) -> Jet<T,N>{y}
fmax(const T& x, const Jet<T, N>& y) -> y
```
This is actually inconsistent with the definition of `std::max()` which returns the *first* value in the case of equality, that is what we used to do, but now we return the second value which is inconsistent.  I don't see any documentation about which value `fmax()` would return in the case of equality, but as it is for simple scalar types for which it would make no difference I don't see why they would feel the need to specify it.  However, `std::max()` *does* as it supports custom comparators and we should revert to be consistent with that.

**Comment by sergiud** [comment]
Thanks for the summary. However, reverting to previous logic does not fix the underlying problem. Always returning a specific argument on equality (a convention) is what I proposed above but does not seem to convince @sandwichmaker.

**Comment by alexsmac** [comment]
I've raised a CL [here](https://ceres-solver-review.googlesource.com/c/ceres-solver/+/21780) that implements the behaviour I described whereby we return the argument with maximal information content, or the first if their information content is equal

**Comment by sandwichmaker** [comment]
The key argument that you make in your CL is that when comparing to a constant we should treat Jets as having more information and that should somehow privilege them. I do not find this argument particularly convincing.  Because mathematically there is no difference between a constant c and a jet with scalar part c and zero derivatives. 

Trying to draw a distinction between the two cases I think will lead us down the wrong path. 

Life outside ceres is also quite confusing.  Here are three derivative computations three version of max.

max(x, x^2), max(x^2, x) and max_reduce[x,x^2]

```
x = tf.Variable(1.0)
with tf.GradientTape() as tape:
  y = tf.math.maximum(x,x**2)
dy=tape.gradient(y,x)

with tf.GradientTape() as tape:
  z = tf.math.maximum(x**2,x)
dz = tape.gradient(z,x)

with tf.GradientTape() as tape:
  zz = tf.math.reduce_max([x,x**2])
dzz = tape.gradient(zz,x)

print(dy.numpy(), dz.numpy(), dzz.numpy())
```

This results in 
```
1.0, 2.0, 1.5
```
because max is sensitive to the ordering, and reduce_max takes the average of its arguments when their values are the same. Jax does averaging too.

So my proposal is to do what Jax is doing, because it does two things.

1. It makes the answer order independent.
2. It handles not just the constant and jet case, but it also gives an answer in the jet and jet case.

**Comment by sandwichmaker** [comment]
Jax does averaging

```
import jax.numpy as jnp
from jax import grad

def myfun(x):
  return jnp.maximum(x, x*x)

print(grad(myfun)(1.0))
```

will print `1.5`.

**Comment by sergiud** [comment]
Averaging makes sense to me and implementation-wise avoids introducing additional overloads to `fmin/fmax`. The hyperplane analogy (assuming the scalar part of a Jet is meant to be the intercept and the derivative is the normal) is not convincing because the arguments cannot be hyperplanes and (incomplete) points at the same time. For a point to be on a hyperplane, the dot product of the normal and the point offset by the intercept must be zero, no?

As for the Jet-Jet case, averaging could only be implemented for the above functions. However, `std::min/std::max` would still return only the first argument on equality. This distinction should be clearly pointed out in the documentation.

**Comment by sandwichmaker** [comment]
@sergiud why are the semantics for std::max and std::fmax different?

**Comment by sandwichmaker** [comment]
Reading 

https://stackoverflow.com/questions/16584558/the-difference-between-max-and-fmax-cross-platform-compiling

I can see that one implementation for std::max is that it always returns the first argument in the case of equality. I think it is extremely unintuitive to be doing this though.

**Comment by sergiud** [comment]
For one, both functions use a single binary comparison. However, we need lexicographical (three-way) comparison. (Fun fact: in C++20  there's a spaceship operator that allows such comparison in a single call.) Also, both functions do not provide means for applying an operation to both arguments post hoc to transform the result.

The C++standard also says:

> Returns the first argument when the arguments are equivalent.

**Comment by sandwichmaker** [comment]
Thanks Sergiu, 
https://pixorblog.wordpress.com/2016/06/27/some-remarks-about-minmax-functions/
has a nice explanation. The short of it is, numerical code should not use std::min/max. we should just mark min/max with warnings.

**Comment by sandwichmaker** [comment]
Now that we have agreement on fmin/fmax, @alexsmac will you revise your Cl to average the gradients.
